File: frontend/src/types/rss.ts
// Types for communication with the API v1 RSS endpoints. Gardarr manages
// qBittorrent's own RSS feeds and auto-downloading rules - it never polls
// feeds or evaluates rules itself, that stays entirely on qBittorrent's side.

export interface RSSArticle {
  id: string;
  title: string;
  summary?: string;
  link?: string;
  is_read: boolean;
}

export interface RSSFeed {
  path: string;
  url?: string;
  title: string;
  last_build?: string;
  is_loading: boolean;
  has_error: boolean;
  articles?: RSSArticle[];
  worker_id?: string;
}

export interface RSSRule {
  name: string;
  enabled: boolean;
  must_contain: string;
  must_not_contain: string;
  use_regex: boolean;
  episode_filter?: string;
  smart_filter: boolean;
  previously_matched_episodes?: string[];
  affected_feeds?: string[];
  ignore_days: number;
  last_match?: string;
  add_paused: boolean;
  assigned_category?: string;
  save_path?: string;
  torrent_content_layout?: string;
  worker_id?: string;
}

export interface RSSAddFeedRequest {
  url: string;
  path?: string;
}

export interface RSSSetFeedURLRequest {
  path: string;
  url: string;
}

export interface RSSAddFolderRequest {
  path: string;
}

export interface RSSMoveItemRequest {
  item_path: string;
  dest_path: string;
}

export interface RSSMarkAsReadRequest {
  item_path: string;
  article_id?: string;
}

export interface RSSSetRuleRequest {
  enabled: boolean;
  must_contain: string;
  must_not_contain: string;
  use_regex: boolean;
  episode_filter?: string;
  smart_filter: boolean;
  affected_feeds: string[];
  ignore_days: number;
  add_paused: boolean;
  assigned_category?: string;
  save_path?: string;
  torrent_content_layout?: string;
}

// Per-worker errors keyed by worker UUID, same convention as the
// aggregated tasks endpoints.
export type RSSWorkerErrors = Record<string, string>;

export interface RSSFeedListResponse {
  feeds: RSSFeed[];
  errors: RSSWorkerErrors;
}

export interface RSSRuleListResponse {
  rules: RSSRule[];
  errors: RSSWorkerErrors;
}

// Matching articles preview: feed name/URL -> matching article titles.
export type RSSMatchingArticles = Record<string, string[]>;

File: backend/internal/entities/rss.go
package entities

import "github.com/google/uuid"

// RSSFeed represents an RSS feed registered on a qBittorrent instance,
// addressed by its Path (folder-qualified, e.g. "Movies\1080p"). qBittorrent
// itself owns polling and article storage; Gardarr only reads and manages
// this state through the WebUI API. WorkerID is set by workermanager.Service
// (zero-valued when a caller already knows the worker, e.g. a single-worker
// repository call) so a feed can be identified in a cross-instance listing.
type RSSFeed struct {
	Path      string
	URL       string
	Title     string
	LastBuild string
	IsLoading bool
	HasError  bool
	Articles  []RSSArticle
	WorkerID  uuid.UUID
}

// RSSArticle is a single item within an RSS feed.
type RSSArticle struct {
	ID      string
	Title   string
	Summary string
	Link    string
	IsRead  bool
}

// RSSRule represents a qBittorrent RSS auto-downloading rule: the actual
// automation engine behind RSS in qBittorrent. Gardarr manages rule
// definitions but never evaluates or triggers them itself - matching and
// downloading stay entirely on the qBittorrent side.
type RSSRule struct {
	Name                      string
	Enabled                   bool
	MustContain               string
	MustNotContain            string
	UseRegex                  bool
	EpisodeFilter             string
	SmartFilter               bool
	PreviouslyMatchedEpisodes []string
	AffectedFeeds             []string
	IgnoreDays                int
	LastMatch                 string
	AddPaused                 bool
	AssignedCategory          string
	SavePath                  string
	TorrentContentLayout      string
	WorkerID                  uuid.UUID
}

// RSSFeedListResult aggregates feeds across every worker, mirroring
// TaskListResult: partial per-worker failures don't fail the whole read.
type RSSFeedListResult struct {
	Feeds  []*RSSFeed
	Errors map[string]string // WorkerUUID -> ErrorMessage
}

// RSSRuleListResult aggregates rules across every worker, mirroring
// TaskListResult.
type RSSRuleListResult struct {
	Rules  []*RSSRule
	Errors map[string]string // WorkerUUID -> ErrorMessage
}
